sdk/src/main/scala/io/horizen/storage/AbstractHistoryStorage.scala

```diff
@@ -139,29 +139,28 @@ abstract class AbstractHistoryStorage[
     sidechainBlockInfo.mainchainHeaderBaseInfo.last
   }
 
-  def getNumOfScBlocksWithNoMcRefDataSince(blockId: ModifierId): Int = {
+  def tooManyBlocksWithoutMcHeadersDataSince(blockId: ModifierId): Boolean = {
     var sidechainBlockInfo: SidechainBlockInfo = this.blockInfoById(blockId)
-    var scBlocksCount : Int= 0
-    var mcRefEmpty : Boolean = sidechainBlockInfo.mainchainHeaderBaseInfo.isEmpty
+    var scBlocksCount : Int = 0
+    var blockHasMcHeaders : Boolean = sidechainBlockInfo.mainchainHeaderBaseInfo.nonEmpty
 
-    // we are going backwards until we find a mc block reference, it should not be possible to
-    // have an infinite loop. But just to be on the safe side we set a limit
-    val safeGuardLimit = params.maxHistoryRewritingLength*100
-
-    while(mcRefEmpty) {
+    // go backwards on the chain until we found a mc header or we hit the genesis block, and break if we
+    // cross the threshold value of max revertable number of blocks
+    while(!blockHasMcHeaders || sidechainBlockInfo.parentId == params.sidechainGenesisBlockId) {
```

Comment:
It is not clear to me this `while` clause, why it should execute the body in case has found a block with headers but its parent is the genesis block? Does this block count as one without header?

Reply:
Fixed, also added a UT for this method